# Introduction

This notebook constitutes the SQL portion of my Practicum Final Project. Here I will be diving in to a database of a startup attempting to create a new app for book lovers following the sudden increase in demand caused by the coronavirus pandemic. 

My goal is to generate a value proposition for a new product. The tasks for this project focus on isolating books for large numbers of high ratings to to determine which publishers and authors are most popular. 

## Initialization

In [ ]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine

In [ ]:
# create database pipeline

db_config = {'user': 'practicum_student',         # username
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [ ]:
# printing top 3 rows from `books` table

query = ''' SELECT *
                FROM books
                LIMIT 3
            '''

pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135

Description of `books`:

- `book_id`
- `author_id`
- `title`
- `num_pages` — number of pages
- `publication_date`
- `publisher_id`

In [ ]:
# checking for duplicate values

query = '''
SELECT book_id
FROM books
GROUP BY book_id
HAVING COUNT(book_id) > 1
'''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [book_id]
Index: []

In [ ]:
# checking for null values

query = '''
SELECT *
FROM books
WHERE coalesce(book_id, author_id, num_pages, publisher_id) IS NULL
'''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [book_id, author_id, title, num_pages, publication_date, publisher_id]
Index: []

In [ ]:
query = '''
SELECT *
FROM books
WHERE title IS NULL
'''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [book_id, author_id, title, num_pages, publication_date, publisher_id]
Index: []

In [ ]:
query = '''
SELECT *
FROM books
WHERE publication_date IS NULL
'''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [book_id, author_id, title, num_pages, publication_date, publisher_id]
Index: []

In [ ]:
# printing top 3 rows from `authors` table

query = ''' SELECT *
                FROM authors
                LIMIT 3
            '''

pd.io.sql.read_sql(query, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie

Description of `authors`:

- `author_id`
- `author`

In [ ]:
# checking for duplicate values

query = '''
SELECT author_id
FROM authors
GROUP BY author_id
HAVING COUNT(author_id) > 1
'''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [author_id]
Index: []

In [ ]:
# checking for null values

query = '''
SELECT *
FROM authors
WHERE author_id IS NULL
'''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [author_id, author]
Index: []

In [ ]:
# printing top 3 rows from `publishers` table

query = ''' SELECT *
                FROM publishers
                LIMIT 3
            '''

pd.io.sql.read_sql(query, con = engine)

publisher_id  publisher
0             1        Ace
1             2   Ace Book
2             3  Ace Books

Description of `publishers`:

- `publisher_id`
- `publisher`

In [ ]:
# checking for duplicate values

query = '''
SELECT publisher_id
FROM publishers
GROUP BY publisher_id
HAVING COUNT(publisher_id) > 1
'''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [publisher_id]
Index: []

In [ ]:
# checking for null values

query = '''
SELECT *
FROM publishers
WHERE publisher_id IS NULL
'''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [publisher_id, publisher]
Index: []

In [ ]:
# printing top 3 rows from `ratings` table

query = ''' SELECT *
                FROM ratings
                LIMIT 3
            '''

pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5

Descriptiono of user `ratings`:

- `rating_id`
- `book_id`
- `username` — the name of the user who rated the book
- `rating`

In [ ]:
# checking for duplicate values

query = '''
SELECT rating_id
FROM ratings
GROUP BY rating_id
HAVING COUNT(rating_id) > 1
'''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [rating_id]
Index: []

In [ ]:
# checking for null values

query = '''
SELECT *
FROM ratings
WHERE coalesce(rating_id,book_id,rating) IS NULL
'''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [rating_id, book_id, username, rating]
Index: []

In [ ]:
# printing top 3 rows from `reviews` table

query = ''' SELECT *
                FROM reviews
                LIMIT 3
            '''

pd.io.sql.read_sql(query, con = engine)

review_id  book_id      username  \
0          1        1  brandtandrea   
1          2        1    ryanfranco   
2          3        2      lorichen   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...

Description of `reviews`:

- `review_id`
- `book_id`
- `username` — the name of the user who reviewed the book
- `text` — the text of the review

In [ ]:
# checking for duplicate values

query = '''
SELECT review_id
FROM reviews
GROUP BY review_id
HAVING COUNT(review_id) > 1
'''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [review_id]
Index: []

In [ ]:
# checking for null values

query = '''
SELECT *
FROM reviews
WHERE coalesce(review_id,book_id) IS NULL
'''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [review_id, book_id, username, text]
Index: []

## Analysis

In [ ]:
# task 1: Find the number of books released after January 1, 2000.

query = ''' 
SELECT COUNT(book_id) 
FROM books b
WHERE publication_date > '2000-01-01'
'''

pd.io.sql.read_sql(query, con = engine)

count
0    819

### Task 1 Notes

Within this database there are 819 books released after Jan 1, 2000. Obviously over 20+ years this number is a tiny fraction of the total number of books released but it will allow for more streamlined analysis. 

In [ ]:
# task 2: Find the number of user reviews and the average rating for each book.

query = '''
SELECT books.title, COUNT (DISTINCT reviews.review_id), AVG(ratings.rating)
FROM books
LEFT JOIN reviews ON books.book_id = reviews.book_id
LEFT JOIN ratings ON reviews.book_id = ratings.book_id
GROUP BY books.title
ORDER BY COUNT(reviews.review_id) DESC
LIMIT 5
'''

pd.io.sql.read_sql(query, con = engine)

title  count       avg
0                            Twilight (Twilight  #1)      7  3.662500
1                The Hobbit  or There and Back Again      6  4.125000
2                             The Catcher in the Rye      6  3.825581
3  Harry Potter and the Prisoner of Azkaban (Harr...      6  4.414634
4  Harry Potter and the Chamber of Secrets (Harry...      6  4.287500

### Task 2 Notes

The book with the most reviews in the database is 'Twilight #1' with 7 unique reviews and an average rating of 3.6/5. The remaining books in the top 5 mpst reviewed books all have 6 revoews. All ratings range between 3.5/5 and 4.5/5.

In [ ]:
# task 3: Identify the publisher that has released the greatest number of books with more than 50 pages 
# (this will help you exclude brochures and similar publications from your analysis).

query = ''' 
SELECT publishers.publisher AS publisher
FROM books 
LEFT JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE books.num_pages > 50
GROUP BY publishers.publisher
ORDER BY COUNT(books.book_id) DESC
LIMIT 10
'''

pd.io.sql.read_sql(query, con = engine)

publisher
0              Penguin Books
1                    Vintage
2   Grand Central Publishing
3           Penguin Classics
4           Ballantine Books
5                     Bantam
6                    Berkley
7         St. Martin's Press
8              Berkley Books
9  William Morrow Paperbacks

### Task 3 Notes

Penguin Books has the most books published with more than 50 pages. The database does seperate publishers by imprints so we can see both Penguin Books and Penguin Classics.|

In [ ]:
# task 4: Identify the author with the highest average book rating: 
# look only at books with at least 50 ratings.

query = '''
SELECT authors.author, AVG(ratings.rating), COUNT(ratings.book_id)
FROM authors
JOIN books ON authors.author_id = books.author_id
JOIN ratings ON books.book_id = ratings.book_id
GROUP BY authors.author
HAVING COUNT(ratings.book_id) > 50
ORDER BY AVG(ratings.rating) DESC
LIMIT 1
'''

pd.io.sql.read_sql(query, con = engine)

author       avg  count
0  J.K. Rowling/Mary GrandPré  4.288462    312

### Task 4 Notes

The author with the highest averge book rating is J.K. Rowling with an average rating 4.29 out of 312 reviews.

In [ ]:
# task 5: Find the average number of text reviews among users who rated more than 50 books.

query = '''
SELECT AVG(count) AS avg_text_reviews
FROM(
    SELECT COUNT(reviews.text)
    FROM(
        SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating) > 50) AS users
    JOIN reviews ON users.username = reviews.username
    GROUP BY users.username) AS count
'''

pd.io.sql.read_sql(query, con = engine)

avg_text_reviews
0         24.333333

### Task 5 Notes

This task required a subquery to be able to find the average of the total number of text reviews made by users with more than 50 ratings. After running this query I found that users who made more than 50 ratings averaged 24.3 text reviews.

## Conclusion

After completing analysis of the data tables I've reached the following conclusions:

- Books with lots of reviews tend to have high ratings with the top 10 books ratings ranging between 3.5/5 and 4.5/5.
- There are a great deal of publishers in the data base which, including imprints of the same publisher (Penguin Books and Penguin Classics). Penguin Books had the most books published in the database. 
- The book with the most reviews is Twilight #1, though this book only has an average rating of 3.6/5. 
- J.K. Rowling, the author with the highest average rating has two books in the top 5 most reviewed books, both with ratings over 4/5.
- Users that write reviews usually write a lot of reviews as the average number of text reviews for users who have published over 50 ratings in 24.3 text reviews. 

My recommendations going forward are the following:

- incentive new users to write reviews. Users that write reviews both read a lot of books and remain very active on the platform.
- Add more books published after Jan 1, 2000 as the database currently only contains 819 books released in the last 20+ years. 